File: Assignment2/WikipediaPopular.java
import java.io.IOException;
import java.util.StringTokenizer;
import org.apache.hadoop.conf.Configuration;
import org.apache.hadoop.conf.Configured;
import org.apache.hadoop.fs.Path;
import org.apache.hadoop.io.IntWritable;
import org.apache.hadoop.io.LongWritable;
import org.apache.hadoop.io.Text;
import org.apache.hadoop.mapreduce.Job;
import org.apache.hadoop.mapreduce.Mapper;
import org.apache.hadoop.mapreduce.Reducer;
import org.apache.hadoop.mapreduce.lib.input.TextInputFormat;
import org.apache.hadoop.mapreduce.lib.output.TextOutputFormat;
import org.apache.hadoop.util.Tool;
import org.apache.hadoop.util.ToolRunner;
import org.apache.hadoop.mapreduce.lib.reduce.LongSumReducer;
import java.lang.*;


public class WikipediaPopular extends Configured implements Tool {

	public static class TokenizerMapper
	extends Mapper<LongWritable, Text, Text, LongWritable>{

		private final static LongWritable count = new LongWritable(0L);
		private Text word = new Text();

		@Override
		public void map(LongWritable key, Text value, Context context
				) throws IOException, InterruptedException {

			StringTokenizer string_arr = new StringTokenizer(value.toString());

			String datetime = string_arr.nextToken();
			String lan = string_arr.nextToken();
			String name = string_arr.nextToken();
			String cnt = string_arr.nextToken();
				
			if(lan.equals("en") && !(name.equals("Main_Page")) && (!name.startsWith("Special:"))){
				word.set(datetime);
				count.set(Long.valueOf(cnt));
				context.write(word,count);
			}

		}
	}


	public static class CountReducer
	extends Reducer<Text, LongWritable, Text, LongWritable> {
		
		private LongWritable result = new LongWritable();

		@Override
		public void reduce(Text key, Iterable<LongWritable> values,
				Context context
				) throws IOException, InterruptedException {
			
			long max_count = 0L;

			for (LongWritable val : values) {
				max_count = Math.max(max_count, Long.valueOf(val.get()));
			}

			result.set(max_count);
			context.write(key, result);
		}
	}


	public static void main(String[] args) throws Exception {
		int res = ToolRunner.run(new Configuration(), new WikipediaPopular(), args);
		System.exit(res);
	}


	@Override
	public int run(String[] args) throws Exception {
		Configuration conf = this.getConf();
		Job job = Job.getInstance(conf, "Wikipedia Popular");
		job.setJarByClass(WikipediaPopular.class);

		job.setInputFormatClass(TextInputFormat.class);

		job.setMapperClass(TokenizerMapper.class);
		job.setReducerClass(CountReducer.class);

		job.setOutputKeyClass(Text.class);
		job.setOutputValueClass(LongWritable.class);
		job.setOutputFormatClass(TextOutputFormat.class);
		TextInputFormat.addInputPath(job, new Path(args[0]));
		TextOutputFormat.setOutputPath(job, new Path(args[1]));

		return job.waitForCompletion(true) ? 0 : 1;
	}
}

File: Assignment2/wikipedia_popular.py
from pyspark import SparkContext, SparkConf
import sys


#input & output
inputs = sys.argv[1]
output = sys.argv[2]

conf = SparkConf().setAppName('Wikipedia_popular')
sc = SparkContext(conf=conf)
assert sys.version_info >= (3, 5)  # make sure we have Python 3.5+
assert sc.version >= '2.3'


def func(line):
	string_arr = line.split()

	datetime = string_arr[0]
	lan = string_arr[1]
	name = string_arr[2]
	cnt = int(string_arr[3])

	return (datetime,lan,name,cnt)


def maximum(x,y):
	if x[0]>=y[0]:
		return x
	else:
		return y


def get_key(kv):
	return kv[0]


def tab_separated(kv):
	k,v = kv
	return "%s\t%s" % (k,v)


text = sc.textFile(inputs)

words = text.map(func)

words2 = words.filter(lambda x: x[1]=="en" and x[2] != "Main_Page" and not (x[2].startswith("Special:")))

pairs = words2.map(lambda x: (x[0],(x[3], x[2])))

redu = pairs.reduceByKey(maximum)

outdata = redu.sortBy(get_key)
outdata.map(tab_separated).saveAsTextFile(output)

print(outdata.take(10))